Added ability to specify penalty factors specific to each coefficient as in glmnet

AsafManela opened the pull request:
Added ability to specify penalty factors specific to each coefficient as in glmnet with corresponding tests added against glmnet results.

Also fixed some bugs with :obj criterion, and added tests (:obj was never tested).

simonster commented:
Thanks! Is there a reason to use a sparse matrix here to back the penalty factors, rather than just using an ordinary Vector? The issue with the sparse matrix is that accessing individual elements is `O(log(n))` whereas an ordinary Vector is `O(1)`, and I don't think the memory savings would be worth that. (A design matrix needs at least one non-zero element per predictor, so the input data size already scales at least with the number of predictors.)

AsafManela commented:
I'm trying to implement a lasso extension that requires an update to the
penalty factors on every CD outer loop iteration, where most of the weights
are 1 and only nonzero coefficients have different weights.
For large p problems I think this SparseWeights type will be faster.
One thing I could do to improve speed in dense problems is to use
SparseWeigths only if they are provided as penalty_weights in fit(), and if
a Vector is provided I'll go with it.
What do you think?

On Wed, Jul 6, 2016 at 5:38 PM, Simon Kornblith <notifications@github.com>
wrote:

> Thanks! Is there a reason to use a sparse matrix here to back the penalty
> factors, rather than just using an ordinary Vector? The issue with the
> sparse matrix is that accessing individual elements is O(log(n)) whereas
> an ordinary Vector is O(1), and I don't think the memory savings would be
> worth that. (A design matrix needs at least one non-zero element per
> predictor, so the input data size already scales at least with the number
> of predictors.)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/simonster/Lasso.jl/pull/6#issuecomment-230928384>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ANLNOun68t_sRt1Je2VE2V9ZXz_-OvHtks5qTC5QgaJpZM4JGjH7>
> .
>

AsafManela commented:
I added several features and fixed a bug or two in LassoPath.
My current master allows for variable specific penalty factors as in glmnet, which as you suggested are now implemented with a dense vector.

I also added:
1. Gadfly-based path plots
2. AICc-based segment selection in coef().
3. GammaLassoPath as a generalization of LassoPath that implements [Taddy (2016)](http://arxiv.org/abs/1308.5623).

All these changes are unit tested and pass for v0.3 and v0.4. Unfortunately, Gadfly currently breaks on nightly, so that one does not pass.

simonster commented:
Hey, sorry I haven't gotten back to you in a long time. I'm busy trying to finish my PhD :smile:. This looks like great work, and I'm going to try to review in detail over the next week.

Re: the regularization path plots, I agree that they're useful, but I'd rather not make this package depend on Gadfly. I think it's probably best to split that functionality into a separate package, and then link to it from the README.

simonster commented on src/Lasso.jl:
```diff
@@ -190,26 +193,35 @@ function Base.show(io::IO, path::LassoPath)
+function computeλ(Xy, λminratio, α, nλ, ω::@compat(Union{Vector,Void}))
     λmax = abs(Xy[1])
-    for i = 1:length(Xy)
+    if ω != nothing
```
It should be a little more efficient if this is written as `!isa(ω, Void)` here and below.

simonster commented on src/Lasso.jl:
```diff
@@ -301,10 +322,191 @@ function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{LassoPath},
+        path.coefs = spzeros(T, p, 0)
+        path.b0 = zeros(T, 0)
+        path.niter = 0
+    end
```
What is the reason for this? If `λ` is cleared it will not be so straightforward to fit the path later.

AsafManela replied:
I had a problem with using code which did not know whether or not the path was fitted, and these
would throw undefined var errors.
I could not figure out a way to test whether it was fitted or not, but if you have better ideas they welcome.

simonster replied:
You could check `isdefined(path, :coefs)`, or else add a new field that specifies whether or not it has been fit.

simonster commented on src/Lasso.jl:
```diff
@@ -301,10 +322,191 @@ function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{LassoPath},
 
+StatsBase.nobs(path::RegularizationPath) = length(path.m.rr.y)
+
+# TODO would be better to use GLM.dispersion_parameter, but that breaks backward comp.
```
No real need to worry about backward compatibility.

simonster commented on src/Lasso.jl:
```diff
@@ -301,10 +322,191 @@ function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{LassoPath},
+    import StatsBase.df, StatsBase.aicc
+end
+
+""" Approximates the degrees-of-freedom in each segment of the path as the number of non zero coefficients
```
The Julia docstring convention is generally to write the function name indented on the first line and leave following lines unindented, e.g. https://github.com/JuliaStats/StatsBase.jl/blob/add47e6b0b44c26540e6bdabd4ce3470582d58ae/src/statmodels.jl

simonster commented on src/Lasso.jl:
```diff
@@ -301,10 +322,191 @@ function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{LassoPath},
+            p+=1
+        end
+        return zeros(eltype(X),p)
+    end
```
If there are no λ values, I would think this function would either return an empty matrix or throw an error.

simonster commented on src/Lasso.jl:
```diff
@@ -301,10 +322,191 @@ function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{LassoPath},
+    else
+        error("unknown selector $select")
+    end
+end
```
This is definitely useful, but it would be good if this function only had one return type, rather than returning either a SparseMatrixCSC or a Vector, because when a function's return type is not determined by its input types it makes Julia's type inference sad. One possibility is to define singleton types for AICc/CVmin/CV1se selection, e.g. `immutable AICc end`, and then have separate methods such as `StatsBase.coef(path::RegularizationPath, ::Type{AICc})`.

simonster replied:
On second thought, maybe `select` is not the proper interface here. If one wants to get the coefficients, predict something, and compute the deviance all at e.g. `CVmin`, it would be better if it weren't necessary to perform cross-validation three times. Perhaps a better approach would be to provide a way to get a model object that represents the model at `CVmin`.

simonster commented on src/Lasso.jl:
```diff
@@ -301,10 +322,191 @@ function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{LassoPath},
+    end
+
+    # calculate etas for each obs x segment
+    eta = newX * coef(path;select=select)
```
Instead of constructing a copy of `newX` augmented by a column of ones and then multiplying it by a copy of the `path.coef` augmented by a row of zeros, it would be more efficient to simply compute `newX * path.coef + path.b0`.

AsafManela replied:
I did that initially, but this way is a bit more elegant using code in coef(). Otherwise, predict would need to redo all select logic. The downside as you note is that it may be less efficient. Do you see a way around it?

simonster replied:
I think the select logic should be separate from `coef`, but even if not, you could always have an argument to `coef` that tells it not to return the intercept.

simonster commented on src/Lasso.jl:
```diff
@@ -301,10 +322,191 @@ function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{LassoPath},
+    else
+        # invert all etas to mus
+        μ(η) = linkinv(mm.rr.l, η)
+        map(μ,eta)
```
Or just `[linkinv(m.rr.l, η) for η in eta]`

AsafManela replied:
Your suggestion returns an array of type Any, which is problematic for users of predict.
Any idea how to overcome that?

simonster replied:
That's a sign that there is a [type stability issue](http://docs.julialang.org/en/release-0.4/manual/performance-tips/#write-type-stable-functions) somewhere, probably with `coef` unless you already changed it. You can use `T[linkinv(m.rr.l, η) for η in eta]` which will always return an array of type T, and on Julia 0.5 you won't even need this change, but it would still be good to fix the type stability issue eventually since it will have a performance impact.

AsafManela replied:
Seems the problem is that eta is a matrix and the comprehension generates a long vector.

simonster replied:
You're right. This has changed in Julia 0.5 (it will be a matrix there), but probably best to stick with what you have for now.

simonster commented on src/Lasso.jl:
```diff
@@ -301,10 +322,191 @@ function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{LassoPath},
+end
+
+"distribution of underlying GLM"
+distfun(path::RegularizationPath) = (typeof(path.m) <: LinearModel) ? Normal() : path.m.rr.d
```
Better to do this by dispatch. If `RegularizationPath` is parametrized by the type of `m`, then you can do:

```julia
distfun{M<:LinearModel}(path::RegularizationPath{M}) = Normal()
distfun(path::RegularizationPath) = path.m.rr.d
```

AsafManela replied:
Can you explain more what you mean by "If RegularizationPath is parametrized by the type of m"?

simonster replied:
Something like:

```julia
abstract RegularizationPath{M<:Union{LinearModel,GeneralizedLinearModel}} <: RegressionModel
type LassoPath{M,T} <: RegularizationPath{M}
(...)
```

simonster commented on src/coordinate_descent.jl:
```diff
@@ -656,7 +677,7 @@ function StatsBase.fit!{S<:GeneralizedLinearModel,T}(path::LassoPath{S,T}; verbo
     if autoλ
         # No need to fit the first model
-        coefs[:, 1] = zero(T)
+        # coefs[:, 1] = zero(T) # redundant
```
Yes, this seems unnecessary. You can just delete the line.

simonster commented on src/coordinate_descent.jl:
```diff
@@ -693,27 +714,35 @@ function StatsBase.fit!{S<:GeneralizedLinearModel,T}(path::LassoPath{S,T}; verbo
+                    for icoef = 1:nnz(newcoef)
+                        oldcoefval = icoef > nnz(oldcoef) ? zero(T) : oldcoef.coef[icoef]
+                        coefdiff.coef[icoef] = newcoef.coef[icoef] - oldcoefval
+                    end
```
Am I missing something, or can this just be:

```julia
for icoef = 1:nnz(oldcoef)
    coefdiff.coef[icoef] -= oldcoef.coef[icoef]
end
```

simonster commented on src/coordinate_descent.jl:
```diff
@@ -693,27 +714,35 @@ function StatsBase.fit!{S<:GeneralizedLinearModel,T}(path::LassoPath{S,T}; verbo
                         for icoef = 1:nnz(newcoef)
                             oldcoefval = icoef > nnz(oldcoef) ? zero(T) : oldcoef.coef[icoef]
-                            newcoef.coef[icoef] = oldcoefval+f*(newcoef.coef[icoef] - oldcoefval)
+                            newcoef.coef[icoef] = oldcoefval+f*(coefdiff.coef[icoef])
```
Thanks for catching this!

simonster commented on src/coordinate_descent.jl:
```diff
@@ -693,27 +714,35 @@ function StatsBase.fit!{S<:GeneralizedLinearModel,T}(path::LassoPath{S,T}; verbo
+                        verbose && println("f=$f: $obj > $objold, dev=$dev, b0=$b0, newcoef=$newcoef")
                         f /= 2.; f > minStepFac || error("step-halving failed at beta = $(newcoef)")
                         for icoef = 1:nnz(newcoef)
                             oldcoefval = icoef > nnz(oldcoef) ? zero(T) : oldcoef.coef[icoef]
```
No longer necessary

simonster commented on src/coordinate_descent.jl:
```diff
@@ -741,11 +770,12 @@ function StatsBase.fit!{S<:GeneralizedLinearModel,T}(path::LassoPath{S,T}; verbo
-            if i == nλ || (autoλ && last_dev_ratio - dev_ratio < MIN_DEV_FRAC_DIFF ||
-                           pct_dev[i] > MAX_DEV_FRAC)
+            if i == nλ || (autoλ && (last_dev_ratio - dev_ratio < MIN_DEV_FRAC_DIFF ||
+                           pct_dev[i] > MAX_DEV_FRAC))
```
Thank you!

simonster commented on src/cross_validation.jl:
```diff
@@ -0,0 +1,188 @@
+#   cvstds = vec(std(oosdevs,2)) / sqrt(nfolds-1)
+#   cv1se = (mincvmean + cvstds[segCVmin]) - cvmeans
+#   segCV1se = minimum(collect(1:size(cvmeans,1))[cv1se.>=0])
+# end
```
Better to either delete this or move it to another file.

simonster commented on src/cross_validation.jl:
```diff
@@ -0,0 +1,188 @@
+
+    CVfun = eval(select)
+    CVfun(oosdevs)
+end
```
Doesn't look used. Should either delete this or move to another file if you think it might be useful later for debugging.

simonster commented on src/cross_validation.jl:
```diff
@@ -0,0 +1,188 @@
+  nλ,nfolds = size(oosdevs)
+  cvmeans = vec(mean(oosdevs,2))
+  (mincvmean,segCVmin) = findmin(cvmeans)
+  mincvstds = std(oosdevs[segCVmin,:]) / sqrt(nfolds-1)
```
It would be marginally efficient to a view here (`view(oosdevs, segCVmin, :)` instead of `oosdevs[segCVmin,:]`), but I doubt this actually a performance bottleneck.

simonster commented on src/cross_validation.jl:
```diff
@@ -0,0 +1,188 @@
+     pathType = GammaLassoPath
+    else
+     error("unknown path typeof(path) $(typeof(path))")
+    end
```
Better to split this up into a separate function:

```julia
pathtype(::LassoPath) = LassoPath
pathtype(::GammaLassoPath) = GammaLassoPath
```

simonster commented on src/cross_validation.jl:
```diff
@@ -0,0 +1,188 @@
+# which chooses λt at lowest mean OOS deviance.
+
+function CVmin(oosdevs)
+  cvmeans = mean(oosdevs,2)
```
Better to use 4 spaces for indentation for consistency.

AsafManela replied:
do you know a good way to auto format a file with mixed tab lengths in Atom?

simonster commented on src/cross_validation.jl:
```diff
@@ -0,0 +1,188 @@
+
+    # EQUAL WEIGHTS ONLY!
+    wts = ones(T, n)
+    devresid_closure(yi, μis) = devresid(d, yi, μis, one(T))
```
I think it's better just to call `devresid` directly below.

simonster commented on src/cross_validation.jl:
```diff
@@ -0,0 +1,188 @@
+      Xwconst = [ones(T,n,1) X]
+    else
+      Xwconst = X
+    end
```
This appears to be unused.

simonster commented on src/cross_validation.jl:
```diff
@@ -0,0 +1,188 @@
+      end
+    end
+
+    CVfun = eval(select)
```
`eval` is evil. It should be possible to avoid here by making the possible `select` arguments types.

simonster commented on src/gammalasso.jl:
```diff
@@ -0,0 +1,406 @@
+    # can potentially pass a different γ for each element of X, but if scalar we copy it to all params
+    p = size(X, 2)
+    if length(γ) == 1
+      γ=convert(Vector{T},repmat([γ],p))
```
If you check that `isa(γ, Number)` then this can just be `fill(T(γ), p)`

simonster commented on src/gammalasso.jl:
```diff
@@ -0,0 +1,406 @@
+    if length(γ) == 1
+      γ=convert(Vector{T},repmat([γ],p))
+    else
+      @assert length(γ)==p "length(γ) != number of parameters ($p)"
```
`@assert` is generally supposed to be used to make sure the programmer didn't make a mistake, not the user. Better to write this as `length(y) == p || throw(DimensionMismatch("length(γ) != number of parameters ($p)")))`

simonster commented on src/gammalasso.jl:
```diff
@@ -0,0 +1,406 @@
+        path.niter = 0
+    end
+    path
+end
```
Is there a way to cut down on the amount of repetition of code from LassoPath here?

simonster commented on src/gammalasso.jl:
```diff
@@ -0,0 +1,406 @@
+    if !isempty(Xnorm)
+        scale!(Xnorm, path.coefs)
+    end
+end
```
Similarly, since these are very similar to `fit` for LassoPath it would be good to avoid repeating all this code.

simonster commented:
Sorry, that was a longer "week" than I imagined. While I've made a bunch of comments, this is great work. See my comment above re: plots, and also, no need to worry about Julia 0.3 support. (Julia 0.5 is on the horizon, and it's no longer possible to register packages that support Julia 0.3 anyway.) Let me know if you have time to work on this further, or if you'd rather that I make the tweaks necessary to satisfy my pedantry.

AsafManela commented:
Thanks for reviewing my code and giving detailed comments. I learned a lot.
I did most of what you suggested, including extracting out the plot functionality to a new LassoPlot.jl package. I won't be able to get to addressing the cross-validation refactoring you suggested, or the item to remove code duplication between LassoPath and GammaLassoPath (which make a lot of sense), at least until early Nov. So if you can take those that would be great.
